File: lib/ingest/claim_mapping.ex
# lib/ingest/claim_mapping.ex — the medipim REFERENCE ADAPTER: envelopes → canonical claims
# (gr-beo; split onto the contract seam in gr-3jd).
#
# Stage 2 of the legacy-medipim ingest, and the adapter every future customer copies: it maps a
# source system's export (here, contract-C HistoryEnvelopes from gr-n8i) to CANONICAL CLAIMS —
# plain wire-shaped maps per docs/CLAIMS_CONTRACT.md (`canonical_claims/1`). The generic half,
# canonical claims → engine claims, lives in `CanonicalClaims` (lib/contract/canonical_claims.ex)
# and is shared with the Product API's live path. `build/1` composes both stages and yields the
# engine's claim log + the `shared` code set, ready for clustering/reconcile (gr-chq:
# `Cluster.variants(Substrate.current(claims), shared)` then `IdentityLedger.decide`).
#
# Being a BACKFILL adapter, its canonical claims carry contract-C unix-second temporal fields and
# `member_of` claims — both deliberately beyond the live wire (see CanonicalClaims' header), so
# build/1 uses the trusted `CanonicalClaims.to_engine!/1` rather than the validating seam.
#
# What the mapping does, per the design (docs/plans/2026-06-05-legacy-history-ingest-design.md):
#
#   1. FOLD per listing = (legacy_entity, source). Replay that source's granular identity events
#      (set/add/remove/delete, in recorded_at order) into a final code-set — a SNAPSHOT (v1).
#      `set` replaces a single-valued scheme (a null value clears it); `add`/`remove` edit a
#      collection; `delete` (op-4) drops the whole scheme entry. Folding runs on medipim's own
#      scheme names so eanGtin13(set) and ean(add) don't interfere, THEN maps to engine schemes.
#
#   2. CANONICALIZE + PARTITION. Every code goes through `Codes.canonicalize` (GTIN family →
#      GTIN-14). Codes that must never bridge two products — restricted/in-store GTINs
#      (`Codes.restricted?`) and non-bridging schemes (MPN/supplier ref) — are collected into the
#      `shared` set the clusterer carries but never fuses on.
#
#   3. BUILD canonical claims (wire-shaped maps; `CanonicalClaims.to_engine!/1` then builds the
#      engine claims via `Substrate.claim/5`, which re-canonicalizes idempotently):
#      * identity  — one per listing: %{"ref" => "entity:source", "codes" => <folded set>}.
#      * grouping  — synthesized, one per (listing, code): %{"code", "product" => legacy_entity}.
#                    Makes the legacy entity a first-class product label for collision reasoning.
#      * attribute — one per attribute event, anchored to the listing's primary code
#                    (CNK ▸ canonical GTIN ▸ first). Survivorship is the engine's job, not ours.
#      * member_of — one per edge add/set, anchored likewise, pointing at the collection.
#
# NOT here: media claims (out of scope for gr-beo), survivorship/clustering (the engine owns it),
# and edge removals (a snapshot-v1 simplification — member_of unions and does not retract).

defmodule ClaimMapping do
  # medipim field → engine scheme atom is owned by CodeRegistry (the single source of medipim code
  # knowledge): the GTIN family all canonicalize to :gtin; each national code keeps its own atom.

  # National SHORT codes, in anchor preference order: a short national code is the most stable
  # primary (cnk for BE, cip_acl7 for FR/LU, …). Preferred over GTINs so Belgian behaviour
  # (CNK-first) is unchanged and French listings anchor on cip_acl7, not a recycled barcode.
  @national_primary [:cnk, :cip_acl7, :cefip, :pzn, :sukl, :pzn_austria, :national_code, :cn]

  # schemes that identify a *supplier's* reference, not a globally-unique product — never bridge.
  @non_bridging_schemes MapSet.new([:mpn, :supplier_ref])

  @doc """
  Map a list of `%HistoryEnvelope{}` to `%{claims: [%Events.ClaimAsserted{}], shared: MapSet}`.
  Claims carry a chronological `order` (later recorded_at ⇒ higher order ⇒ wins survivorship).

  Two stages: `canonical_claims/1` (this adapter) then `CanonicalClaims.to_engine!/1` (the
  generic contract seam) — `to_engine!` because the backfill flavor deliberately exceeds the
  live-wire validator (member_of + unix-second temporals; see CanonicalClaims).
  """
  def build(envelopes) when is_list(envelopes) do
    folded = fold_raw(envelopes)

    %{
      claims: envelopes |> canonical(folded) |> CanonicalClaims.to_engine!() |> stamp(),
      shared: folded |> listing_codes() |> shared_codes()
    }
  end

  @doc """
  Stage (a) alone: the canonical claims this adapter derives from the envelopes — wire-shaped
  maps per docs/CLAIMS_CONTRACT.md, in emission order (identity, grouping, attribute,
  member_of), each carrying contract-C unix-second `"valid_from"`/`"recorded_at"`. This is what
  a customer's mapping script would produce and submit.
  """
  def canonical_claims(envelopes) when is_list(envelopes),
    do: canonical(envelopes, fold_raw(envelopes))

  defp canonical(envelopes, folded) do
    listing_codes = listing_codes(folded)
    entity_codes = union_by_entity(listing_codes)

    listing_primary = Map.new(listing_codes, fn {k, set} -> {k, primary(Enum.sort(set))} end)
    entity_primary = Map.new(entity_codes, fn {e, set} -> {e, primary(Enum.sort(set))} end)

    # A sourced event anchors ONLY to its own listing's primary code; if that listing folded to
    # empty (delisted), the event is skipped rather than re-homed onto another source's code.
    # Only genuinely unsourced events fall back to the entity-level primary.
    anchor = fn
      entity, nil -> Map.get(entity_primary, entity)
      entity, source -> Map.get(listing_primary, {entity, source})
    end

    # Deterministic emission order — Map/MapSet iteration order is otherwise unspecified, which
    # would let stamp/1's tie-break (and primary/1's pick among equals) drift between runs/versions.
    ordered = Enum.sort_by(listing_codes, fn {k, _set} -> k end)

    identity =
      for {{e, s} = k, set} <- ordered do
        %{
          "kind" => "identity",
          "source" => s,
          "ref" => "#{e}:#{s}",
          "codes" => set |> Enum.sort() |> Enum.map(&CanonicalClaims.code_string/1),
          "valid_from" => folded[k].last_at,
          "recorded_at" => folded[k].last_at
        }
      end

    grouping =
      for {{e, s} = k, set} <- ordered, code <- Enum.sort(set) do
        %{
          "kind" => "grouping",
          "source" => s,
          "code" => CanonicalClaims.code_string(code),
          "product" => e,
          "valid_from" => folded[k].last_at,
          "recorded_at" => folded[k].last_at
        }
      end

    attribute =
      for env <- envelopes,
          ev <- env.events,
          ev.kind == :attribute,
          a = anchor.(env.legacy_entity, ev.source),
          a != nil do
        %{
          "kind" => "attribute",
          "source" => ev.source,
          "code" => CanonicalClaims.code_string(a),
          "field" => field_dim(ev),
          "value" => ev.data.value,
          "valid_from" => ev.valid_from,
          "recorded_at" => ev.recorded_at
        }
      end

    member_of =
      for env <- envelopes,
          ev <- env.events,
          ev.kind == :edge,
          ev.op in [:set, :add],
          ev.data.value != nil,
          a = anchor.(env.legacy_entity, ev.source),
          a != nil do
        %{
          "kind" => "member_of",
          "source" => ev.source,
          "code" => CanonicalClaims.code_string(a),
          "collection" => ev.data.collection,
          "member" => to_string(ev.data.value),
          "valid_from" => ev.valid_from,
          "recorded_at" => ev.recorded_at
        }
      end

    identity ++ grouping ++ attribute ++ member_of
  end

  @doc "Just the folded, canonicalized code-set per listing — `%{{entity, source} => MapSet}`."
  def listings(envelopes) when is_list(envelopes), do: envelopes |> fold_raw() |> listing_codes()

  # Per-listing canonicalized code-sets, with delisted (now-empty) listings dropped.
  defp listing_codes(folded) do
    folded
    |> Map.new(fn {k, v} -> {k, engine_codes(v.raw)} end)
    |> Enum.reject(fn {_k, set} -> MapSet.size(set) == 0 end)
    |> Map.new()
  end

  # ── fold ──────────────────────────────────────────────────────────────────

  # Replay identity events into per-listing raw code-sets, keyed by medipim scheme name.
  # Envelope events are already time-ordered, so list order is recorded_at order.
  defp fold_raw(envelopes) do
    for env <- envelopes, ev <- env.events, ev.kind == :identity, reduce: %{} do
      acc ->
        key = {env.legacy_entity, ev.source}
        cur = Map.get(acc, key, %{raw: %{}, last_at: 0})
        Map.put(acc, key, %{raw: apply_identity(cur.raw, ev), last_at: max(cur.last_at, ev.recorded_at)})
    end
  end

  @doc false
  # Shared with FinerClaims (the per-event fold) — same delta semantics, one implementation.
  def apply_identity(raw, ev) do
    scheme = ev.data.scheme
    code = ev.data.code

    case ev.op do
      :set when is_nil(code) -> Map.delete(raw, scheme)
      :set -> Map.put(raw, scheme, MapSet.new([code]))
      :add -> Map.update(raw, scheme, MapSet.new([code]), &MapSet.put(&1, code))
      :remove -> raw |> Map.update(scheme, MapSet.new(), &MapSet.delete(&1, code)) |> drop_empty(scheme)
      :delete -> Map.delete(raw, scheme)
    end
  end

  defp drop_empty(raw, scheme) do
    case Map.get(raw, scheme) do
      %MapSet{} = s -> if MapSet.size(s) == 0, do: Map.delete(raw, scheme), else: raw
      _ -> raw
    end
  end

  # raw (medipim scheme → values) → MapSet of canonicalized engine codes.
  @doc false
  # Shared with FinerClaims.
  def engine_codes(raw) do
    for {scheme, values} <- raw, v <- values, into: MapSet.new() do
      Codes.canonicalize({scheme_atom(scheme), v})
    end
  end

  # Known medipim fields map to engine atoms via the registry (the GTIN family → :gtin); an
  # unrecognised field stays the raw string (Codes.canonicalize passes it through). The registry
  # never String.to_atom/1's an unknown field — the loader does not whitelist schemes, so that
  # would be an atom-table leak.
  defp scheme_atom(scheme), do: CodeRegistry.scheme(scheme)

  # ── helpers ───────────────────────────────────────────────────────────────

  defp union_by_entity(listing_codes) do
    Enum.reduce(listing_codes, %{}, fn {{e, _s}, set}, acc ->
      Map.update(acc, e, set, &MapSet.union(&1, set))
    end)
  end

  # primary code for anchoring: a national SHORT code (in @national_primary order) ▸ non-restricted
  # canonical GTIN ▸ any GTIN ▸ a 13-digit national code (acl13/cip13) ▸ lowest code. National
  # short codes win so Belgian listings still anchor on CNK and French ones on cip_acl7 (a stable
  # id) rather than a recycled barcode.
  @doc false
  # Shared with FinerClaims — anchoring must pick the same primary in both folds.
  def primary([]), do: nil

  def primary(codes) do
    national_short(codes) ||
      Enum.find(codes, &(match?({:gtin, _}, &1) and not Codes.restricted?(&1))) ||
      Enum.find(codes, &match?({:gtin, _}, &1)) ||
      Enum.find(codes, &match?({:acl13, _}, &1)) ||
      Enum.find(codes, &match?({:cip13, _}, &1)) ||
      codes |> Enum.sort() |> List.first()
  end

  # First code whose scheme appears earliest in @national_primary (cnk ▸ cip_acl7 ▸ …).
  defp national_short(codes) do
    Enum.find_value(@national_primary, fn scheme ->
      Enum.find(codes, &match?({^scheme, _}, &1))
    end)
  end

  @doc false
  # Shared with FinerClaims.
  def field_dim(ev) do
    case ev.data.locale do
      nil -> ev.data.field
      locale -> "#{ev.data.field}:#{locale}"
    end
  end

  defp shared_codes(listing_codes) do
    for {_k, set} <- listing_codes, code <- set, shared?(code), into: MapSet.new(), do: code
  end

  @doc false
  # Shared with FinerClaims — both folds must agree on what may never bridge.
  def shared?({scheme, _} = code),
    do: Codes.restricted?(code) or MapSet.member?(@non_bridging_schemes, scheme)

  # chronological order: later recorded_at ⇒ higher order. Stable on the original emission index.
  defp stamp(claims) do
    claims
    |> Enum.with_index()
    |> Enum.sort_by(fn {c, i} -> {c.recorded_at, i} end)
    |> Enum.with_index()
    |> Enum.map(fn {{c, _i}, order} -> %{c | order: order} end)
  end
end

File: php/src/ClaimMapping.php
<?php

declare(strict_types=1);

namespace GoldenRecord;

/**
 * The medipim reference adapter — ported from `ClaimMapping` (lib/ingest/claim_mapping.ex).
 *
 * Folds contract-C HistoryEnvelopes into canonical claims (`canonicalClaims`) and composes them
 * into engine claims plus the `shared` code set (`build`). Per listing = (legacy_entity, source):
 * replay identity events into a final code-set, canonicalize/partition, then synthesize identity /
 * grouping / attribute / member_of claims, plus first-class lane records (media + descriptions) tied
 * back by depicts/describes edges. Claims are stamped with a chronological `order`.
 */
final class ClaimMapping
{
    /** National short codes, in anchor preference order. */
    private const NATIONAL_PRIMARY = ['cnk', 'cip_acl7', 'cefip', 'pzn', 'sukl', 'pzn_austria', 'national_code', 'cn'];

    private const NON_BRIDGING_SCHEMES = ['mpn', 'supplier_ref'];

    /** medipim edge collections that reference first-class entities. collection => [scheme, lane, relation]. */
    private const LANE_COLLECTIONS = [
        'descriptions' => ['text_id', 'description', 'describes'],
        'media' => ['asset_id', 'media', 'depicts'],
    ];

    /**
     * Map envelopes to ['claims' => [ClaimAsserted...], 'shared' => code-set].
     *
     * @param list<array<string,mixed>> $envelopes
     * @return array{claims: list<array<string,mixed>>, shared: array<string, array{0: string, 1: string}>}
     */
    public static function build(array $envelopes): array
    {
        $folded = self::foldRaw($envelopes);

        $canonical = self::canonical($envelopes, $folded);
        $claims = self::stamp(CanonicalClaims::toEngineBang($canonical));
        $shared = self::sharedCodes(self::listingCodes($folded));

        return ['claims' => $claims, 'shared' => $shared];
    }

    /**
     * Stage (a): the canonical claims (wire-shaped maps) in emission order.
     *
     * @param list<array<string,mixed>> $envelopes
     * @return list<array<string,mixed>>
     */
    public static function canonicalClaims(array $envelopes): array
    {
        return self::canonical($envelopes, self::foldRaw($envelopes));
    }

    /**
     * Just the folded, canonicalized code-set per listing — keyed "entity\x1fsource".
     *
     * @param list<array<string,mixed>> $envelopes
     * @return array<string, array<string, array{0: string, 1: string}>>
     */
    public static function listings(array $envelopes): array
    {
        return self::listingCodes(self::foldRaw($envelopes));
    }

    /**
     * @param list<array<string,mixed>> $envelopes
     * @param array<string, array{raw: array<string, array<string,true>>, last_at: int}> $folded
     * @return list<array<string,mixed>>
     */
    private static function canonical(array $envelopes, array $folded): array
    {
        $listingCodes = self::listingCodes($folded); // key => code-set
        $entityCodes = self::unionByEntity($listingCodes); // entity => code-set

        // primary code per listing / per entity.
        $listingPrimary = [];
        foreach ($listingCodes as $k => $set) {
            $listingPrimary[$k] = self::primary(Sets::valuesSorted($set));
        }
        $entityPrimary = [];
        foreach ($entityCodes as $e => $set) {
            $entityPrimary[(string) $e] = self::primary(Sets::valuesSorted($set));
        }

        // Deterministic emission order: sort listings by their key.
        $orderedKeys = array_keys($listingCodes);
        sort($orderedKeys, SORT_STRING);

        $identity = [];
        $grouping = [];
        foreach ($orderedKeys as $key) {
            [$entity, $source] = self::splitKey($key);
            $set = $listingCodes[$key];
            $sortedCodes = Sets::valuesSorted($set);

            $identity[] = [
                'kind' => 'identity',
                'source' => $source,
                'ref' => $entity.':'.$source,
                'codes' => array_map(CanonicalClaims::codeString(...), $sortedCodes),
                'valid_from' => $folded[$key]['last_at'],
                'recorded_at' => $folded[$key]['last_at'],
            ];

            foreach ($sortedCodes as $code) {
                $grouping[] = [
                    'kind' => 'grouping',
                    'source' => $source,
                    'code' => CanonicalClaims::codeString($code),
                    'product' => self::entityValue($entity),
                    'valid_from' => $folded[$key]['last_at'],
                    'recorded_at' => $folded[$key]['last_at'],
                ];
            }
        }

        // anchor(entity, source): a sourced event anchors only to its OWN listing's primary; an
        // unsourced event falls back to the entity-level primary.
        $anchor = static function (mixed $entity, ?string $source) use ($listingPrimary, $entityPrimary): ?array {
            if ($source === null) {
                return $entityPrimary[(string) $entity] ?? null;
            }

            return $listingPrimary[self::makeKey($entity, $source)] ?? null;
        };

        $attribute = [];
        foreach ($envelopes as $env) {
            foreach ($env['events'] as $ev) {
                if ($ev['kind'] !== 'attribute') {
                    continue;
                }
                $a = $anchor($env['legacy_entity'], $ev['source']);
                if ($a === null) {
                    continue;
                }
                $attribute[] = [
                    'kind' => 'attribute',
                    'source' => $ev['source'],
                    'code' => CanonicalClaims::codeString($a),
                    'field' => self::fieldDim($ev),
                    'value' => $ev['data']['value'],
                    'valid_from' => $ev['valid_from'],
                    'recorded_at' => $ev['recorded_at'],
                ];
            }
        }

        $memberOf = [];
        foreach ($envelopes as $env) {
            foreach ($env['events'] as $ev) {
                if ($ev['kind'] !== 'edge') {
                    continue;
                }
                if (!in_array($ev['op'], ['set', 'add'], true)) {
                    continue;
                }
                if (($ev['data']['value'] ?? null) === null) {
                    continue;
                }
                if (isset(self::LANE_COLLECTIONS[$ev['data']['collection']])) {
                    continue;
                }
                $a = $anchor($env['legacy_entity'], $ev['source']);
                if ($a === null) {
                    continue;
                }
                $memberOf[] = [
                    'kind' => 'member_of',
                    'source' => $ev['source'],
                    'code' => CanonicalClaims::codeString($a),
                    'collection' => $ev['data']['collection'],
                    'member' => self::toStringValue($ev['data']['value']),
                    'valid_from' => $ev['valid_from'],
                    'recorded_at' => $ev['recorded_at'],
                ];
            }
        }

        $laneEntities = self::laneEntities($envelopes, $anchor);

        return array_merge($identity, $grouping, $attribute, $memberOf, $laneEntities);
    }

    /**
     * First-class lane records (media + descriptions): fold per (entity, source, collection), emit
     * an identity claim in the entity's lane + a typed edge back to the listing's anchor.
     *
     * @param list<array<string,mixed>> $envelopes
     * @param callable(mixed, ?string): ?array{0: string, 1: string} $anchor
     * @return list<array<string,mixed>>
     */
    private static function laneEntities(array $envelopes, callable $anchor): array
    {
        $refs = self::laneRefs($envelopes); // key "entity\x1fsource\x1fcollection" => {ids, last, entity, source, collection}

        $keys = array_keys($refs);
        sort($keys, SORT_STRING);

        $out = [];
        foreach ($keys as $key) {
            $ref = $refs[$key];
            [$scheme, $lane, $relation] = self::LANE_COLLECTIONS[$ref['collection']];

            $a = $anchor($ref['entity'], null);
            if ($a === null) {
                continue;
            }

            $ids = array_keys($ref['ids']);
            sort($ids, SORT_STRING);
            foreach ($ids as $id) {
                [$vf, $at] = $ref['last'][$id];
                $out[] = [
                    'kind' => 'identity',
                    'source' => $ref['source'],
                    'ref' => $ref['collection'].':'.$id,
                    'codes' => [$scheme.':'.$id],
                    'entity' => $lane,
                    'valid_from' => $vf ?? $at,
                    'recorded_at' => $at,
                ];
                $out[] = [
                    'kind' => 'edge',
                    'source' => $ref['source'],
                    'from' => $scheme.':'.$id,
                    'relation' => $relation,
                    'to' => CanonicalClaims::codeString($a),
                    'valid_from' => $vf ?? $at,
                    'recorded_at' => $at,
                ];
            }
        }

        return $out;
    }

    /**
     * Fold media events per (entity, source||source_system, collection): add/remove churn on the
     * asset id, so only surviving references remain.
     *
     * @param list<array<string,mixed>> $envelopes
     * @return array<string, array{ids: array<string,true>, last: array<string, array{0: mixed, 1: mixed}>, entity: mixed, source: string, collection: string}>
     */
    private static function laneRefs(array $envelopes): array
    {
        $acc = [];
        foreach ($envelopes as $env) {
            foreach ($env['events'] as $ev) {
                if ($ev['kind'] !== 'media') {
                    continue;
                }
                if (!isset(self::LANE_COLLECTIONS[$ev['data']['collection']])) {
                    continue;
                }
                $source = $ev['source'] ?? $env['source_system'];
                $key = self::makeKey($env['legacy_entity'], $source)."\x1f".$ev['data']['collection'];
                $id = self::toStringValue($ev['data']['asset']);

                if (!isset($acc[$key])) {
                    $acc[$key] = [
                        'ids' => [],
                        'last' => [],
                        'entity' => $env['legacy_entity'],
                        'source' => $source,
                        'collection' => $ev['data']['collection'],
                    ];
                }
                if ($ev['op'] === 'remove') {
                    unset($acc[$key]['ids'][$id]);
                } else {
                    $acc[$key]['ids'][$id] = true;
                }
                $acc[$key]['last'][$id] = [$ev['valid_from'], $ev['recorded_at']];
            }
        }

        return $acc;
    }

    // ── fold ────────────────────────────────────────────────────────────────────

    /**
     * Replay identity events into per-listing raw code-sets, keyed by medipim scheme name.
     *
     * @param list<array<string,mixed>> $envelopes
     * @return array<string, array{raw: array<string, array<string,true>>, last_at: int}>
     */
    private static function foldRaw(array $envelopes): array
    {
        $acc = [];
        foreach ($envelopes as $env) {
            foreach ($env['events'] as $ev) {
                if ($ev['kind'] !== 'identity') {
                    continue;
                }
                $key = self::makeKey($env['legacy_entity'], $ev['source']);
                if (!isset($acc[$key])) {
                    $acc[$key] = ['raw' => [], 'last_at' => 0];
                }
                $acc[$key]['raw'] = self::applyIdentity($acc[$key]['raw'], $ev);
                $acc[$key]['last_at'] = max($acc[$key]['last_at'], $ev['recorded_at']);
            }
        }

        return $acc;
    }

    /**
     * Apply one identity delta (set/add/remove/delete) to a raw code-set (scheme => set-of-values).
     *
     * @param array<string, array<string,true>> $raw
     * @param array<string,mixed> $ev
     * @return array<string, array<string,true>>
     */
    private static function applyIdentity(array $raw, array $ev): array
    {
        $scheme = $ev['data']['scheme'];
        $code = $ev['data']['code'] ?? null;

        switch ($ev['op']) {
            case 'set':
                if ($code === null) {
                    unset($raw[$scheme]);
                } else {
                    $raw[$scheme] = [(string) $code => true];
                }

                return $raw;
            case 'add':
                $raw[$scheme][(string) $code] = true;

                return $raw;
            case 'remove':
                if (isset($raw[$scheme])) {
                    unset($raw[$scheme][(string) $code]);
                    if ($raw[$scheme] === []) {
                        unset($raw[$scheme]);
                    }
                }

                return $raw;
            case 'delete':
                unset($raw[$scheme]);

                return $raw;
            default:
                return $raw;
        }
    }

    /**
     * raw (medipim scheme → values) → code-set of canonicalized engine codes.
     *
     * @param array<string, array<string,true>> $raw
     * @return array<string, array{0: string, 1: string}>
     */
    private static function engineCodes(array $raw): array
    {
        $set = [];
        foreach ($raw as $scheme => $values) {
            foreach (array_keys($values) as $v) {
                $code = Codes::canonicalize([CodeRegistry::scheme($scheme), (string) $v]);
                $set[Codes::key($code)] = $code;
            }
        }

        return $set;
    }

    /**
     * Per-listing canonicalized code-sets, with delisted (now-empty) listings dropped.
     *
     * @param array<string, array{raw: array<string, array<string,true>>, last_at: int}> $folded
     * @return array<string, array<string, array{0: string, 1: string}>>
     */
    private static function listingCodes(array $folded): array
    {
        $out = [];
        foreach ($folded as $key => $v) {
            $set = self::engineCodes($v['raw']);
            if ($set !== []) {
                $out[$key] = $set;
            }
        }

        return $out;
    }

    // ── helpers ──────────────────────────────────────────────────────────────────

    /**
     * @param array<string, array<string, array{0: string, 1: string}>> $listingCodes
     * @return array<string, array<string, array{0: string, 1: string}>> entity => code-set
     */
    private static function unionByEntity(array $listingCodes): array
    {
        $acc = [];
        foreach ($listingCodes as $key => $set) {
            [$entity] = self::splitKey($key);
            $acc[$entity] = isset($acc[$entity]) ? Sets::union($acc[$entity], $set) : $set;
        }

        return $acc;
    }

    /**
     * Primary anchor code: national short code ▸ non-restricted GTIN ▸ any GTIN ▸ acl13 ▸ cip13 ▸
     * lowest code.
     *
     * @param list<array{0: string, 1: string}> $codes already sorted by [scheme, value]
     * @return array{0: string, 1: string}|null
     */
    public static function primary(array $codes): ?array
    {
        if ($codes === []) {
            return null;
        }

        $national = self::nationalShort($codes);
        if ($national !== null) {
            return $national;
        }

        foreach ($codes as $c) {
            if ($c[0] === 'gtin' && !Codes::restricted($c)) {
                return $c;
            }
        }
        foreach ($codes as $c) {
            if ($c[0] === 'gtin') {
                return $c;
            }
        }
        foreach ($codes as $c) {
            if ($c[0] === 'acl13') {
                return $c;
            }
        }
        foreach ($codes as $c) {
            if ($c[0] === 'cip13') {
                return $c;
            }
        }

        $sorted = $codes;
        usort($sorted, Sets::compareCodes(...));

        return $sorted[0];
    }

    /**
     * @param list<array{0: string, 1: string}> $codes
     * @return array{0: string, 1: string}|null
     */
    private static function nationalShort(array $codes): ?array
    {
        foreach (self::NATIONAL_PRIMARY as $scheme) {
            foreach ($codes as $c) {
                if ($c[0] === $scheme) {
                    return $c;
                }
            }
        }

        return null;
    }

    /** @param array<string,mixed> $ev */
    public static function fieldDim(array $ev): string
    {
        $locale = $ev['data']['locale'] ?? null;

        return $locale === null ? $ev['data']['field'] : $ev['data']['field'].':'.$locale;
    }

    /**
     * @param array<string, array<string, array{0: string, 1: string}>> $listingCodes
     * @return array<string, array{0: string, 1: string}>
     */
    private static function sharedCodes(array $listingCodes): array
    {
        $out = [];
        foreach ($listingCodes as $set) {
            foreach (Sets::values($set) as $code) {
                if (self::isShared($code)) {
                    $out[Codes::key($code)] = $code;
                }
            }
        }

        return $out;
    }

    /** @param array{0: string, 1: string} $code */
    public static function isShared(array $code): bool
    {
        return Codes::restricted($code) || in_array($code[0], self::NON_BRIDGING_SCHEMES, true);
    }

    /**
     * Chronological order stamp: later recorded_at ⇒ higher order; stable on emission index.
     *
     * @param list<array<string,mixed>> $claims
     * @return list<array<string,mixed>>
     */
    private static function stamp(array $claims): array
    {
        $indexed = [];
        foreach ($claims as $i => $c) {
            $indexed[] = [$c['recorded_at'], $i, $c];
        }
        usort($indexed, static function (array $a, array $b): int {
            return [self::numeric($a[0]), $a[1]] <=> [self::numeric($b[0]), $b[1]];
        });

        $out = [];
        foreach ($indexed as $order => [$_at, $_i, $c]) {
            $c['order'] = $order;
            $out[] = $c;
        }

        return $out;
    }

    private static function numeric(mixed $v): int|float
    {
        return is_numeric($v) ? 0 + $v : 0;
    }

    private static function makeKey(mixed $entity, string $source): string
    {
        return self::entityKeyPart($entity)."\x1f".$source;
    }

    /** @return array{0: mixed, 1: string} */
    private static function splitKey(string $key): array
    {
        $parts = explode("\x1f", $key, 2);

        return [self::decodeEntity($parts[0]), $parts[1] ?? ''];
    }

    private static function entityKeyPart(mixed $entity): string
    {
        // Encode the entity scalar so an int 1 and string "1" can't collide in the key.
        return (is_int($entity) ? 'i:' : 's:').$entity;
    }

    private static function decodeEntity(string $part): mixed
    {
        if (str_starts_with($part, 'i:')) {
            return (int) substr($part, 2);
        }

        return substr($part, 2);
    }

    private static function entityValue(mixed $entity): mixed
    {
        return $entity;
    }

    /**
     * Mirror Elixir's `to_string/1` over the JSON-decoded values that appear as edge `value`s.
     * A list is an iolist/charlist: integers are bytes (codepoints), strings/lists concatenate.
     * Such a value is then canonicalized downstream (which trims it), e.g. `[9]` → "\t" → "".
     */
    private static function toStringValue(mixed $v): string
    {
        if (is_bool($v)) {
            return $v ? 'true' : 'false';
        }
        if (is_array($v)) {
            $out = '';
            foreach ($v as $part) {
                $out .= is_int($part) ? mb_chr($part, 'UTF-8') : self::toStringValue($part);
            }

            return $out;
        }

        return (string) $v;
    }
}
